### PR TITLE
Implement REST API (parse JSON request)

### PR DESCRIPTION
**Pull request by Leowario**

### CODE REVIEW

**Comment by Leowario** [comment]
@internal-jew PTAL

File: rest-api/src/main/java/com/alevel/deliverit/EndPoint.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.alevel.deliverit.customers.Parcel;
+import com.alevel.deliverit.customers.ParcelReception;
+import com.alevel.deliverit.customers.Sender;
+import org.json.simple.JSONObject;
+
+/**
+ * @author Vitalii Usatyi
+ */
+public class EndPoint {
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/EndPoint.java]: It's just `Endpoint`

File: rest-api/src/main/java/com/alevel/deliverit/ObjectParser.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * @author Vitalii Usatyi
+ */
+public class ObjectParser {
+    private Parcel parcel;
+    private Sender sender;
+
+    /**
+     * Parse JSON request to java objects: {@link Parcel} and {@link Sender}
+     * @param jsonObject - is a JSON request
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/ObjectParser.java]: `@param` doesn't require `-`

File: rest-api/src/main/java/com/alevel/deliverit/ParcelFactory.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+import static com.alevel.deliverit.logistics.WeightUnit.KILOGRAM;
+import static com.alevel.deliverit.logistics.WeightUnit.POUND;
+
+/**
+ * @author Vitalii Usatyi
+ */
+class ParcelFactory {
+
+    Parcel create(JSONObject jsonObject){
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/ParcelFactory.java]: It's useful to develop a habit of formatting code before every commit.

File: rest-api/src/main/java/com/alevel/deliverit/ParcelFactory.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        Map parcelMap = (Map) jsonObject.get("parcel");
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/ParcelFactory.java]: Was there any way to do it automatically?

File: rest-api/src/main/java/com/alevel/deliverit/ObjectParser.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    private Parcel parcel;
+    private Sender sender;
+
+    /**
+     * Parse JSON request to java objects: {@link Parcel} and {@link Sender}
+     *
+     * @param jsonObject is a JSON request
+     */
+    void parseToObject(JSONObject jsonObject) {
+        sender = new SenderFactory().create(jsonObject);
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/ObjectParser.java]: If you decide to use ad-hoc approach to parse entities and value objects, it makes sense to put specific parsers inside of the classes they parse.

Comment (internal-jew) [reply rest-api/src/main/java/com/alevel/deliverit/ObjectParser.java]:
```java
class Parcel {
  ...
  public static Parser parser() {return new ParcelParser();}

  public static class ParcelParser implements Parser { }
}
```

File: rest-api/src/main/java/com/alevel/deliverit/Endpoint.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * @author Vitalii Usatyi
+ */
+public class Endpoint {
+    private ObjectParser objectParser = new ObjectParser();
+    private JsonParser jsonParser = new JsonParser();
+
+    void run(JSONObject jsonObject) {
+        objectParser.parseToObject(jsonObject);
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/Endpoint.java]: Parser inputParser = ParserFactory.getParser(Sender.class);

File: rest-api/src/main/java/com/alevel/deliverit/Endpoint.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import org.json.simple.JSONObject;
+
+/**
+ * @author Vitalii Usatyi
+ */
+public class Endpoint {
+    private ObjectParser objectParser = new ObjectParser();
+    private JsonParser jsonParser = new JsonParser();
+
+    void run(JSONObject jsonObject) {
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/Endpoint.java]: We can use here a `String` instead of `JSONObject`

Comment (internal-jew) [reply rest-api/src/main/java/com/alevel/deliverit/Endpoint.java]: an thus hide details about a library we use.

File: rest-api/src/main/java/com/alevel/deliverit/ParcelFactory.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    private ParcelId buildParcelId(long parcelId) {
+        return new ParcelId(parcelId);
+    }
+
+    private WeightUnit getWeightUnit(Map parcelMap) {
+        String weightUnit = (String) parcelMap.get("weightUnit");
+        if (weightUnit.equals("kilogram")) {
+            return KILOGRAM;
+        }
+        if (weightUnit.equals("pound")) {
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/ParcelFactory.java]: 1. Extend WeightUnit with value for string representation.
2. Compare against this value. `WeightUnit.KILOGRAM.value.equals(weightUnit)`

File: rest-api/src/main/java/com/alevel/deliverit/SenderFactory.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import com.alevel.deliverit.logistics.Country;
+import com.alevel.deliverit.logistics.PostalAddress;
+import org.json.simple.JSONObject;
+
+import java.util.Map;
+
+/**
+ * @author Vitalii Usatyi
+ */
+class SenderFactory {
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/SenderFactory.java]: If we call it a `Parser` then we can accept `Strings`, parse them and return a ready-to-use object.

**Comment by Leowario** [comment]
@internal-jew PTAL

**Comment by Leowario** [comment]
@internal-jew PTAL

File: model/src/main/java/com/alevel/deliverit/customers/Parcel.java [hunk 10]
```diff
@@ -11,7 +11,11 @@
     private final Weight weight;
     private final PostalAddress deliveryAddress;
 
-    protected Parcel(ParcelId id, Weight weight, PostalAddress deliveryAddress) {
+    public static Parser<Parcel> parser() {
```
Comment (internal-jew) [inline model/src/main/java/com/alevel/deliverit/customers/Parcel.java]: Put it under ctor, please.

File: model/src/main/java/com/alevel/deliverit/customers/ParcelParser.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        long id = (long) parcelMap.get("parcelId");
+        ParcelId parcelId = buildParcelId(id);
+        Weight weight = buildWeight(weightDouble, weightUnit);
+        String postalAddressString = parcelMap.get("deliveryAddress").toString();
+        PostalAddress deliveryAddress = PostalAddressParser.parser().parse(postalAddressString);
+        return new Parcel(parcelId, weight, deliveryAddress);
+    }
+
+    private double getWeightDouble(Map parcelMap) {
+        return Double.parseDouble((String) parcelMap.get("weight"));
```
Comment (internal-jew) [inline model/src/main/java/com/alevel/deliverit/customers/ParcelParser.java]: static import

File: model/src/main/java/com/alevel/deliverit/customers/ParcelParser.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    private ParcelId buildParcelId(long parcelId) {
+        return new ParcelId(parcelId);
+    }
+
+    private WeightUnit getWeightUnit(Map parcelMap) {
+
+        String weightUnit = (String) parcelMap.get("weightUnit");
+        if (WeightUnit.KILOGRAM.toString().equalsIgnoreCase(weightUnit)) {
```
Comment (internal-jew) [inline model/src/main/java/com/alevel/deliverit/customers/ParcelParser.java]: static import of `KILOGRAM`

Comment (internal-jew) [reply model/src/main/java/com/alevel/deliverit/customers/ParcelParser.java]: And please add `label` to `WeightUnit` and compare against it

File: model/src/main/java/com/alevel/deliverit/customers/ParcelParser.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            return KILOGRAM;
+        }
+        if (WeightUnit.POUND.toString().equalsIgnoreCase(weightUnit)) {
+            return POUND;
+        } else {
+            throw new IllegalArgumentException("Uncorrected weight unit");
```
Comment (internal-jew) [inline model/src/main/java/com/alevel/deliverit/customers/ParcelParser.java]: Invalid

File: model/src/main/java/com/alevel/deliverit/customers/Parser.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
+package com.alevel.deliverit.customers;
+
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+/**
+ * @author Vitalii Usatyi
```
Comment (internal-jew) [inline model/src/main/java/com/alevel/deliverit/customers/Parser.java]: Please add some documentation

File: model/src/main/java/com/alevel/deliverit/customers/Parser.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
+package com.alevel.deliverit.customers;
```
Comment (internal-jew) [inline model/src/main/java/com/alevel/deliverit/customers/Parser.java]: This class belongs to `customers` package. Rather `com.alevel.deliverit` and `com.alevel.deliverit.model`

File: model/src/main/java/com/alevel/deliverit/customers/Parser.java [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+ */
+public abstract class Parser<T> {
+
+    private T t;
+
+    public T parse(String json) {
+        return t;
+    }
+
+    JSONObject parseToJsonObject(String jsonString) {
```
Comment (internal-jew) [inline model/src/main/java/com/alevel/deliverit/customers/Parser.java]: `protected`

File: model/src/main/java/com/alevel/deliverit/customers/PostalAddressParser.java [hunk 17]
```diff
@@ -0,0 +1,36 @@
+package com.alevel.deliverit.customers;
+
+import com.alevel.deliverit.logistics.Country;
+import com.alevel.deliverit.logistics.PostalAddress;
```
Comment (internal-jew) [inline model/src/main/java/com/alevel/deliverit/customers/PostalAddressParser.java]: It seems that `PostalAddressParser` should be in the same package as `PostalAddress`

File: model/src/main/java/com/alevel/deliverit/customers/SenderParser.java [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.Map;
+
+/**
+ * @author Vitalii Usatyi
+ */
+class SenderParser extends Parser<Sender> {
+
+    @Override
+    public Sender parse(String jsonString) {
+
```
Comment (internal-jew) [inline model/src/main/java/com/alevel/deliverit/customers/SenderParser.java]: redundant line

File: model/src/main/java/com/alevel/deliverit/customers/SenderProfileParser.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+public class SenderProfileParser extends Parser<SenderProfile> {
+
+    private SenderProfileParser() {
+
+    }
+
+    static SenderProfileParser parser() {
+        return new SenderProfileParser();
+    }
+
```
Comment (internal-jew) [inline model/src/main/java/com/alevel/deliverit/customers/SenderProfileParser.java]: redundant line

File: rest-api/src/main/java/com/alevel/deliverit/Endpoint.java [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+        ParcelReceptionRequest request = ParcelReceptionRequest.parser().parse(jsonString);
+        ParcelReception parcelReception =
+                ParcelReception
+                        .builder()
+                        .setSender(request.getSender())
+                        .setParcel(request.getParcel())
+                        .build();
+        parcelReception.accept();
+    }
+
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/Endpoint.java]: Redundant empty lines

File: rest-api/src/test/java/com/alevel/deliverit/RestTest.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+@DisplayName("Rest API should:")
+class RestTest {
+
+    @Test
+    @DisplayName("Parse json to Object")
+    void testParseToObject() {
+//        ObjectParser objectParser = new ObjectParser();
+        Endpoint endpoint = new Endpoint();
+        try {
+            File file = new File("E:\\Java\\Java_test\\JsonRequest.json");
```
Comment (internal-jew) [inline rest-api/src/test/java/com/alevel/deliverit/RestTest.java]: Please, remove from the changeset

**Comment by Leowario** [comment]
@internal-jew PTAL

**Comment by Leowario** [comment]
@internal-jew PTAL

File: rest-api/src/main/java/com/alevel/deliverit/RestVerticle.java [hunk 22]
```diff
@@ -1,53 +1,39 @@
...
 import io.vertx.core.http.HttpMethod;
 import io.vertx.core.http.HttpServer;
-import io.vertx.core.http.HttpServerRequest;
 import io.vertx.core.http.HttpServerResponse;
-import io.vertx.ext.web.Router;
+
 
 /**
- * Receipt HTTP POST request and call EndPoint.start(...) .
+ * Receipt HTTP POST request{@link ParcelReceptionEndpoint#accept(String)}.
```
Comment (internal-jew) [inline rest-api/src/main/java/com/alevel/deliverit/RestVerticle.java]: `request {@link`

**Pull request merged**